Classes/Hooks/HookProvider.php

```diff
@@ -0,0 +1,149 @@
+<?php
+declare(strict_types=1);
+
+namespace OliverKlee\Seminars\Hooks;
+
+use OliverKlee\Seminars\Hooks\Interfaces\Hook;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+
+/**
+ * This class provides functions for unified hooks.
+ *
+ * A hook allows to add functionality at certain points of the program path. These points are
+ * grouped using an interface, declaring the methods to implement to hook in.
+ *
+ * Hooking in
+ * The FQCN of the interface is the key to `$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars']`
+ * where you register your hooked-in classes. Your class will be instantiated once when the first
+ * point is reached and re-used for all other points.
+ *
+ * Implementing hook points
+ * Instantiate this class with the interface you need implemented. First call to `executeHook()` will
+ * instantiate the registered classes. Every further call will reuse the same instances. On each
+ * call provide the method required at the point in your program.
+ *
+ * There is an optional index to `$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars']`, provided
+ * for easier conversion of existing hooks to this class.
+ *
+ * @author Michael Kramer <m.kramer@mxp.de>
+ */
+class HookProvider
+{
+    /**
+     * @var string
+     */
+    protected $interfaceName = '';
+
+    /**
+     * Index in `$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars']` of hooked-in classes
+     *
+     * @var string
+     */
+    protected $index = '';
+
+    /**
+     * @var Hook[]
+     */
+    protected $hookObjects = [];
+
+    /**
+     * @var bool
+     */
+    protected $hooksHaveBeenRetrieved = false;
+
+    /**
+     * @param string $interfaceName interface the hook needs implemented
+     * @param string $index index to `$GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars']`
+     *               if not using the interface name (for backwards compatibility)
+     *               (the interface name is recommended)
+     *
+     * @throws \UnexpectedValueException
+     */
+    public function __construct(string $interfaceName, string $index = '')
+    {
+        if (!\interface_exists($interfaceName)) {
+            throw new \UnexpectedValueException(
+                'The interface ' . $interfaceName . ' does not exist.',
+                1565089078
+            );
+        }
+        if (!\in_array(Hook::class, \class_implements($interfaceName), true)) {
+            throw new \UnexpectedValueException(
+                'The interface ' . $interfaceName . ' does not extend ' . Hook::class . ' interface.',
+                1565088963
+            );
+        }
+
+        $this->interfaceName = $interfaceName;
+        $this->index = $index === '' ? $interfaceName : $index;
+    }
+
+    /**
+     * Execute the hooked-in methods
```

Comment:
In general, please use the third person for the first sentence of method comments: (This method …) "Executes the hooked-in methods."